crates/ruff_python_codegen/src/generator.rs

```diff
@@ -146,20 +145,55 @@ impl<'a> Generator<'a> {
         self.p(s.as_str());
     }
 
-    fn p_bytes_repr(&mut self, s: &[u8], quote: Quote) {
-        let escape = AsciiEscape::with_preferred_quote(s, quote);
+    fn p_bytes_repr(&mut self, s: &[u8], flags: BytesLiteralFlags) {
+        // raw bytes are interpreted without escapes and should all be ascii, but double check
+        // before writing to self
+        if flags.prefix().is_raw() && self.p_raw_bytes(s, flags).is_ok() {
+            return;
+        }
+        let escape = AsciiEscape::with_preferred_quote(s, flags.quote_style());
         if let Some(len) = escape.layout().len {
             self.buffer.reserve(len);
         }
-        escape.bytes_repr().write(&mut self.buffer).unwrap(); // write to string doesn't fail
+        escape
+            .bytes_repr(flags.is_triple_quoted())
+            .write(&mut self.buffer)
+            .unwrap(); // write to string doesn't fail
     }
 
-    fn p_str_repr(&mut self, s: &str, quote: Quote) {
-        let escape = UnicodeEscape::with_preferred_quote(s, quote);
+    /// Returns a [`std::str::Utf8Error`] if `s` is not valid UTF-8, otherwise converts `s` to a
+    /// `str` and adds it to `self`.
+    fn p_raw_bytes(
+        &mut self,
+        s: &[u8],
+        flags: BytesLiteralFlags,
+    ) -> Result<(), std::str::Utf8Error> {
+        let body = std::str::from_utf8(s)?;
+        self.p(flags.prefix().as_str());
+        self.p(flags.quote_str());
+        self.p(body);
+        self.p(flags.quote_str());
+        Ok(())
+    }
+
+    fn p_str_repr(&mut self, s: &str, flags: impl Into<AnyStringFlags>) {
+        let flags = flags.into();
+        if flags.prefix().is_raw() {
+            self.p(flags.prefix().as_str());
+            self.p(flags.quote_str());
+            self.p(s);
+            self.p(flags.quote_str());
```

Comment:
Strictly speaking, this might be slightly less performant, since `format_string_contents` allocates a new `String`. I think it's worth it to avoid repetition, however. The `Generator` also shouldn't be _terribly_ performance-sensitive, since we only create autofixes if we found an error in the user's code that leads us to emit a diagnostic, _and_ we deem that the diagnostic is autofixable.

```suggestion
            self.p(&flags.format_string_contents(s));
```

Reply:
Wow thank you! I must have scrolled past this method so many times in the past few days!

Reply:
Hehe, it's on the `StringFlags` trait that `StringLiteralFlags`, `BytesLiteralFlags`, `FStringFlags` and `AnyStringFlags` all implement :-)